petastorm/pytorch.py

```diff
@@ -367,3 +367,123 @@ def __enter__(self):
+            _sanitize_pytorch_types(row_as_dict)
+
+            for k, v in row_as_dict.items():
+                if not self.reader.batched_output:
+                    row_as_dict[k] = self.transform_fn([v])
+                else:
+                    row_as_dict[k] = self.transform_fn(v)
+
+            if not self.keys:
+                self.keys = row_as_dict.keys()
+
+            # Add rows to buffer
+            items = list(row_as_dict.values())
+            expected_rows = n_rows + len(items[0])
+            last_row = len(items[0])
+
+            if self.capacity <= expected_rows:
+                buffer_full = True
+                last_row = self.capacity-n_rows
+                expected_rows = self.capacity
+            if self.buffer is None:
+                # Initialize buffer as a list of empty tensors
+                self.buffer = []
+                for v in items:
+                    self.buffer.append(torch.empty((self.capacity,) + v.shape[1:], dtype=v.dtype, device=v.device))
+            # Copy new items into buffer
+            for k, v in enumerate(items):
+                self.buffer[k][n_rows:expected_rows] = v[:last_row]
+            n_rows = expected_rows
+
+        # At this point, dataloader has enough rows storted in memory.
+        # Set last_row_consumed = True in reader rather than draining the remainder
+        # of reader. If reader has infinite epochs, draining will be deadlock.
+        for row in self.reader:
+            self.reader.last_row_consumed = True
+        # Truncate empty tensors if capacity is larger than total rows.
+        if n_rows < self.capacity:
+            for k, v in enumerate(self.buffer):
+                self.buffer[k] = self.buffer[k][:n_rows]
+
+    def _iter_impl(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        self._load_rows_into_mem()
+
+        for epoch in range(self.num_epochs):
+            self.size = len(self.buffer[0])
+            if self.shuffle:
+                # Deterministically shuffle based on seed and current epoch id.
+                g = torch.Generator()
+                g.manual_seed(self.seed + epoch)
+                indices = torch.randperm(self.size, generator=g).tolist()
+            else:
+                indices = list(range(self.size))
+
+            # Sample batches
+            for i in range(0, len(indices), self.batch_size):
+                idx = indices[i:i+self.batch_size]
+                batch = [v[idx] for v in self.buffer]
+                self.size -= len(batch[0])
+                batch = dict(zip(self.keys, batch))
+                yield batch
+            assert self.size == 0
+
+    # Functions needed to treat data loader as a context manager
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.reader.stop()
```

Comment:
Your logic shuts the reader earlier - once all data was read. It's a good thing since we release resources we would no longer need. No need to `stop()` the reader here, but ok to join to make sure the `InMemBatchedDataLoader` indeed acts as  a context-manager and does not exit until everything is clean.